crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,261 @@
+/// | ID: 1v1  | PointLight | RenderEntity(ID: 3V1) | SyncRenderWorld |
+/// | ID: 18v1 | PointLight | RenderEntity(ID: 5V1) | SyncRenderWorld |
+/// |-----------------------------------------------------------------|
+///
+/// |----------Render World-----------|
+/// |  Entity  |       Component      |
+/// |---------------------------------|
+/// | ID: 3v1  | MainEntity(ID: 1V1)  |
+/// | ID: 5v1  | MainEntity(ID: 18V1) |
+/// |---------------------------------|
+///
+/// ```
+///
+/// Note that not this effectively establishes a link between the main world entity and the render world entity.
+/// Not every entity needs to be synchronized, however, only entities with [`SyncRenderWorld`] are synced.
+/// Adding [`SyncRenderWorld`] to a main world component will establish such a link.
+/// Once a synchronized main entity is despawned, it's corresponding render entity will be automatically
+/// despawned in the next `sync`.
+///
+/// The sync step does not handle the transfer of component data between worlds,
+/// since it's often not necessary to transfer over all the components of a main world entity.
+/// The render world probably cares about a `Position` component, but not a `Velocity` component.
+/// The extraction happens in it's own step, independently from synchronization.
+///
+/// Moreover, [`WorldSyncPlugin`] only synchronizes *entities*, stuff like mesh and texture data is handled
+/// differently, as those assets are extracted to render world resources and not entities.
+///
+/// [`PipelinedRenderingPlugin`]: crate::pipelined_rendering::PipelinedRenderingPlugin
+#[derive(Default)]
+pub struct WorldSyncPlugin;
+
+impl Plugin for WorldSyncPlugin {
+    fn build(&self, app: &mut bevy_app::App) {
+        app.init_resource::<PendingSyncEntity>();
+        app.observe(
+            |trigger: Trigger<OnAdd, SyncRenderWorld>, mut pending: ResMut<PendingSyncEntity>| {
+                pending.push(EntityRecord::Added(trigger.entity()));
+            },
+        );
+        app.observe(
+            |trigger: Trigger<OnRemove, SyncRenderWorld>,
+             mut pending: ResMut<PendingSyncEntity>,
+             query: Query<&RenderEntity>| {
+                if let Ok(e) = query.get(trigger.entity()) {
+                    pending.push(EntityRecord::Removed(e.id()));
+                };
+            },
+        );
+    }
+}
+/// Marker component that indicates that its entity needs to be Synchronized to the render world
+///
+/// NOTE: This component should persist throughout the entity's entire lifecycle.
+/// If this component is removed from its entity, the entity will be despawned.
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[reflect[Component]]
+#[component(storage = "SparseSet")]
+pub struct SyncRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug, Copy)]
+/// Marker component added on the main world entities that are synced to the Render World in order to keep track of the corresponding render world entity
```

Comment:
Doc comments should go above the derive, not below.

Also do we want to make a note to use required components in the future to avoid the archetype move on insert, or?